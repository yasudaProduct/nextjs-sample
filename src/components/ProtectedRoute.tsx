import React ,{ useEffect } from "react";
import { useSession } from "next-auth/react";
import { useRouter } from "next/router";

const ProtectedRoute = ({ children }: {children: React.ReactNode}) => {

    const router = useRouter();

    const { status } = useSession();

    useEffect(() => {
        if (status === 'unauthenticated') {
          router.push('/api/auth/signin' + '?callbackUrl=' + router.asPath);
        }
      }, [router, status]);

      if(status === 'unauthenticated') return null;

      return <>{children}</>;
}

export default ProtectedRoute;