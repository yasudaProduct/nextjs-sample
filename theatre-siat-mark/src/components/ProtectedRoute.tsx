import React ,{ useEffect } from "react";
import { useSession } from "next-auth/react";
import { useRouter } from "next/router";



const ProtectedRoute = ({ children }: {children: React.ReactNode}) => {

    const router = useRouter();

    const { status } = useSession();

    useEffect(() => {
        if (status === 'unauthenticated') {
          router.push('/auth/login');
        }
      }, [router, status]); //実行タイミング ページもしくはstatusが変更された時

      if(status === 'unauthenticated') return null;

      return <>{children}</>;
}

export default ProtectedRoute;