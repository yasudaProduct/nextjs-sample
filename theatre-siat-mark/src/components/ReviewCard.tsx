import React from "react"
import { useState } from "react"
import { Card, CardContent } from "./ui/card"
import { Bookmark, Star } from "lucide-react"
import { Button } from "./ui/button"

export interface ReviewCardProps {
    review: {
      id: number
      user: { name: string }
      seatNumber: string
      rating: number
      review: string
      isBookmarked?: boolean
    }
}

export function ReviewCard({ review }: ReviewCardProps) {
    const [isBookmarked, setIsBookmarked] = useState(review.isBookmarked)

    const handleBookmark = async () => {
    
        const method = isBookmarked ? 'DELETE' : 'POST'
        const url = isBookmarked 
          ? `/api/bookmarks?reviewId=${review.id}`
          : '/api/bookmarks'
        try {
          const response = await fetch(url, {
            method,
            headers: {
              'Content-Type': 'application/json',
            },
            body: method === 'POST' ? JSON.stringify({ reviewId: review.id }) : undefined,
          })
    
          if (response.ok) {
            setIsBookmarked(!isBookmarked)
          }
        } catch (error) {
          console.error('ブックマークの更新に失敗しました', error)
        }
      }

    return (
        <Card data-testid="review-card">
          <CardContent className="p-4">
            <div className="flex justify-between items-start">
              <div>
                <p className="font-semibold">{review.user.name}</p>
                <p className="text-sm text-gray-500">座席: {review.seatNumber}</p>
              </div>
              <div className="flex items-center space-x-2">
                <div className="flex">
                  {[...Array(5)].map((_, i) => (
                    <Star
                      key={i}
                      className={`h-5 w-5 ${
                        i < review.rating ? 'text-yellow-400 fill-current' : 'text-gray-300'
                      }`}
                    />
                  ))}
                </div>
                <Button
                  variant="ghost"
                  size="icon"
                  onClick={handleBookmark}
                  aria-label={isBookmarked ? "ブックマークを解除" : "ブックマークに追加"}
                >
                  <Bookmark className={isBookmarked ? "text-blue-500 fill-current" : "text-gray-500"} />
                </Button>
              </div>
            </div>
            <p className="mt-2">{review.review}</p>
          </CardContent>
        </Card>
      )
}