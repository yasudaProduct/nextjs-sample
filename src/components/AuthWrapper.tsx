import React from "react";
import { useSession } from "next-auth/react";
import { useRouter } from "next/router";
import ProtectedRoute from "./ProtectedRoute";

const authRoutes = [
  '/mypage',
  '/reviews/my-reviews',
  '/reviews/register-review',
  '/articles/[id]'
];

const AuthWrapper = ({ children }: { children: React.ReactNode }) => {
  const session = useSession();
  const router = useRouter();

  // 認証中は表示しない
  if (session.status === 'loading') return null;

  return (
    <>
      {authRoutes.includes(router.pathname) ? (
        // 現在のページが、ログインを要求するページの場合
        <ProtectedRoute>{children}</ProtectedRoute>
      ) : (
        // 現在のページが、ログインを要求しないページの場合
        children
      )}
    </>
  );

};

export default AuthWrapper;