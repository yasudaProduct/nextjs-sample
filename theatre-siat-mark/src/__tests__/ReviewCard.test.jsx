import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { ReviewCard } from '../components/ReviewCard'

describe('Page', () => {
  it('renders ReviewCard component', () => {
    render(<ReviewCard />);
    const reviewCardElement = screen.getByTestId('review-card');
    expect(reviewCardElement).toBeInTheDocument();
  });

  it('displays correct props', () => {

    const review = {
        id: 1,
        user: { name: "yuta" },
        seatNumber: "D15",
        rating: 1,
        review: "good",
        isBookmarked: false,
    }

    render(<ReviewCard review={review} />)

    const reviewCardElement = screen.getByTestId('review-card')
    expect(reviewCardElement).toBeInTheDocument()

    expect(screen.getByText('yuta')).toBeInTheDocument();
    expect(screen.getByText((content, element) => content.includes('D15'))).toBeInTheDocument();
    expect(screen.getByText('good')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const review = {
      id: 1,
      user: { name: "yuta" },
      seatNumber: "D15",
      rating: 1,
      review: "good",
      isBookmarked: false,
  }
    const { asFragment } = render(<ReviewCard review={review} />);
    expect(asFragment()).toMatchSnapshot();
  });
})