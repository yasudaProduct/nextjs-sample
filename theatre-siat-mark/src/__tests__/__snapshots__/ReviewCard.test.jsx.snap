// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm"
    data-testid="review-card"
  >
    <div
      class="p-4"
    >
      <div
        class="flex justify-between items-start"
      >
        <div>
          <p
            class="font-semibold"
          >
            yuta
          </p>
          <p
            class="text-sm text-gray-500"
          >
            座席: D15
          </p>
        </div>
        <div
          class="flex items-center space-x-2"
        >
          <div
            class="flex"
          >
            <svg
              class="lucide lucide-star h-5 w-5 text-yellow-400 fill-current"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
              />
            </svg>
            <svg
              class="lucide lucide-star h-5 w-5 text-gray-300"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
              />
            </svg>
            <svg
              class="lucide lucide-star h-5 w-5 text-gray-300"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
              />
            </svg>
            <svg
              class="lucide lucide-star h-5 w-5 text-gray-300"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
              />
            </svg>
            <svg
              class="lucide lucide-star h-5 w-5 text-gray-300"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
              />
            </svg>
          </div>
          <button
            aria-label="ブックマークに追加"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10"
          >
            <svg
              class="lucide lucide-bookmark text-gray-500"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m19 21-7-4-7 4V5a2 2 0 0 1 2-2h10a2 2 0 0 1 2 2v16z"
              />
            </svg>
          </button>
        </div>
      </div>
      <p
        class="mt-2"
      >
        good
      </p>
    </div>
  </div>
</DocumentFragment>
`;
